### PR TITLE
menu: add menu generation

### PR DESCRIPTION
**Pull request by Garfield-fr**
This feature allows flexibility to generate menus by programming. The developer has the possibility to extend these functions by adding extensions for processing options. He can also develop his own display by generating a new component.

* Adds a factory for creating menus.
* Adds MenuItem class allowing to define the options of a menu line.
* Adds CoreExtension class to handle basic options.
* Refactors ng-core-tester to use the new menu generation.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Why are you opening this PR?

- a new method to generate menu

## How to test?

- show menu example and declaration.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: projects/ng-core-tester/src/app/app.component.ts [hunk 1]
```diff
@@ -19,6 +19,8 @@ import { TranslateService } from '@ngx-translate/core';
 import { CoreConfigService, RecordEvent, RecordService, TitleMetaService } from '@rero/ng-core';
 import { BsLocaleService } from 'ngx-bootstrap/datepicker';
 import { ToastrService } from 'ngx-toastr';
+import { AppMenuService } from './service/app-menu.service';
+import { MenuItem } from 'projects/rero/ng-core/src/lib/menu/menu-item';
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/app.component.ts:23]: Should be imported from `@rero/ng-core`.

File: projects/ng-core-tester/src/app/home/home.component.ts [hunk 2]
```diff
@@ -14,17 +14,19 @@
...
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
 import { ApiService, DialogService, RecordSearchService, TranslateLanguageService } from '@rero/ng-core';
 import { ToastrService } from 'ngx-toastr';
 import { DocumentComponent } from '../record/document/document.component';
 import { NgxSpinnerService } from 'ngx-spinner';
+import { MenuService } from '../service/menu.service';
+import { MenuItem } from 'projects/rero/ng-core/src/lib/menu/menu-item';
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/home/home.component.ts:23]: Should be imported from `@rero/ng-core`.

File: projects/ng-core-tester/src/app/home/home.component.ts [hunk 3]
```diff
@@ -61,26 +63,23 @@ export class HomeComponent {
...
-      }
-    ]
-  };
+  // Menu
+  menuApplication: MenuItem;
+
+  // Menu click
+  menuClick: any;
+
+  ngOnInit() {
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/home/home.component.ts:72]: Missing doc block.

File: projects/ng-core-tester/src/app/service/menu.service.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { Injectable } from '@angular/core';
+import { MenuFactory } from 'projects/rero/ng-core/src/lib/menu/menu-factory';
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/service/menu.service.ts:18]: Should be imported from `@rero/ng-core`

File: projects/ng-core-tester/src/app/service/menu.service.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      .setUri('http://sonar.ch')
+      .setAttributes({ id: 'sonar-link', target: '_blank' })
+      .setExtra('iconClass', 'fa fa-external-link');
+
+    this._externalLinks(menu);
+
+    return menu;
+  }
+
+  /** Sub Menu Records */
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/service/menu.service.ts:65]: Missing doc for the parameter.

File: projects/ng-core-tester/src/app/service/menu.service.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      .setRouterLink(['/unisi', 'record', 'search', 'documents']);
+    records.addChild('Backend records')
+      .setRouterLink(['/admin', 'record', 'search', 'documents']);
+    records.addChild('Document records')
+      .setRouterLink(['/records', 'documents']);
+    records.addChild('Organisation records')
+      .setRouterLink(['/records', 'organisations']);
+  }
+
+  private _externalLinks(menu: any) {
```
Comment (sebastiendeleze) [inline projects/ng-core-tester/src/app/service/menu.service.ts:82]: Same here.

File: projects/rero/ng-core/src/lib/menu/extension/options-interface.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+export interface OptionsInterface {
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/menu/extension/options-interface.ts:17]: Maybe a short description of what is the purpose of this interface.

**Review by sebastiendeleze — CHANGES_REQUESTED** [review]

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

File: projects/ng-core-tester/src/app/service/app-menu.service.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import { Injectable } from '@angular/core';
+import { MenuFactory } from 'projects/rero/ng-core/src/lib/menu/menu-factory';
+import { TranslateService } from '@ngx-translate/core';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AppMenuService {
+
+  private _factory: MenuFactory;
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/app-menu.service.ts:26]: done

File: projects/ng-core-tester/src/app/service/app-menu.service.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+  constructor(private _translate: TranslateService) {
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/app-menu.service.ts:28]: done

File: projects/ng-core-tester/src/app/service/app-menu.service.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { Injectable } from '@angular/core';
+import { MenuFactory } from 'projects/rero/ng-core/src/lib/menu/menu-factory';
+import { TranslateService } from '@ngx-translate/core';
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/app-menu.service.ts:19]: done

File: projects/ng-core-tester/src/app/app.component.ts (hunk 1 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/app.component.ts:23]: done

File: projects/ng-core-tester/src/app/home/home.component.ts (hunk 3 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/home/home.component.ts:72]: done

File: projects/ng-core-tester/src/app/service/menu.service.ts (hunk 6 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/menu.service.ts:82]: done

File: projects/ng-core-tester/src/app/service/menu.service.ts (hunk 5 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/menu.service.ts:65]: done

File: projects/ng-core-tester/src/app/home/home.component.ts (hunk 2 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/home/home.component.ts:23]: done

File: projects/ng-core-tester/src/app/service/menu.service.ts (hunk 4 above)
Comment (Garfield-fr) [reply projects/ng-core-tester/src/app/service/menu.service.ts:18]: done

File: projects/rero/ng-core/src/lib/menu/extension/options-interface.ts (hunk 7 above)
Comment (Garfield-fr) [reply projects/rero/ng-core/src/lib/menu/extension/options-interface.ts:17]: done

**Review by zannkukai — APPROVED** [review]

**Pull request merged**